### PR TITLE
Environments in the C target

### PR DESCRIPTION
**Pull request by erlingrj**
This PR is (mostly) about introducing the environment concept in reactor-c. There are also changes related to enclaves (mostly contained in `CEnclavedReactorTransformation.java`). This PR is mainly about:

1. Code-generating some functions related to creating and initializing the environments of the program
2. Move global variables tracking things like, all the timers, all startup reactions, shutdown reactions, etc onto the environment.
3. Changes in code-generation of `_lf_initialize_trigger_objects` so that we incoroprate the move from global variables mentioned in (2) to variables on the environment struct.

Companion reactor-c PR: https://github.com/lf-lang/reactor-c/pull/212

### CODE REVIEW

**Review by edwardalee — APPROVED** [review]
Oops, my mistake!  Your fix looks right! Thanks!

File: core/src/main/java/org/lflang/generator/c/CReactionGenerator.java
Comment (erlingrj) [inline core/src/main/java/org/lflang/generator/c/CReactionGenerator.java:330]: Remove

File: core/src/main/java/org/lflang/generator/c/CEnvironmentFunctionGenerator.java
Comment (erlingrj) [inline core/src/main/java/org/lflang/generator/c/CEnvironmentFunctionGenerator.java:1]: Docs

File: core/src/main/java/org/lflang/generator/c/CReactionGenerator.java
Comment (erlingrj) [inline core/src/main/java/org/lflang/generator/c/CReactionGenerator.java:371]: remove fixme

File: core/src/main/java/org/lflang/generator/c/CReactionGenerator.java
Comment (erlingrj) [inline core/src/main/java/org/lflang/generator/c/CReactionGenerator.java:589]: remove fixme

File: core/src/main/java/org/lflang/generator/c/CUtil.java
Comment (erlingrj) [inline core/src/main/java/org/lflang/generator/c/CUtil.java:850]: I think they should be removed as they are unused?

File: core/src/main/java/org/lflang/generator/python/PythonGenerator.java
Comment (erlingrj) [inline core/src/main/java/org/lflang/generator/python/PythonGenerator.java:114]: Remove fixme

File: core/src/testFixtures/java/org/lflang/tests/Configurators.java
Comment (erlingrj) [inline core/src/testFixtures/java/org/lflang/tests/Configurators.java:73]: Which zephyr emulation should we use? Wasnt there  a memory size issue with m3?

File: test/C/src/concurrent/TimeoutZeroThreaded.lf
Comment (erlingrj) [inline test/C/src/concurrent/TimeoutZeroThreaded.lf:9]: remove

File: .github/workflows/c-zephyr-tests.yml
Comment (lhstrh) [inline .github/workflows/c-zephyr-tests.yml:44]: What is the reason for factoring this out into smaller steps?

File: .gitignore
Comment (lhstrh) [inline .gitignore:164]:
```suggestion
.west/

```

File: core/src/main/java/org/lflang/AttributeUtils.java
Comment (lhstrh) [inline core/src/main/java/org/lflang/AttributeUtils.java:251]:
```suggestion
   * Annotate @{code node} with enclave @attribute
```

File: core/src/main/java/org/lflang/generator/EnclaveInfo.java
Comment (lhstrh) [inline core/src/main/java/org/lflang/generator/EnclaveInfo.java:3]: Would it be better to use a `record` here (instead of a `class`)?

File: core/src/main/java/org/lflang/generator/c/CEnclavedReactorTransformation.java
Comment (lhstrh) [inline core/src/main/java/org/lflang/generator/c/CEnclavedReactorTransformation.java:1]: This class needs some JavaDoc...

File: core/src/main/java/org/lflang/generator/c/CReactionGenerator.java
Comment (lhstrh) [reply core/src/main/java/org/lflang/generator/c/CReactionGenerator.java:330]: ?

**Review by lhstrh — CHANGES_REQUESTED** [review]
This is a really impressive effort, @erlingrj! I only left some basic code hygiene remarks. I noticed there are quite a few `FIXME` lingering in the code. I know this might be an annoying request, but if possible, please link them to a GitHub issue for tracking...

**Comment by erlingrj** [comment]
> This is a really impressive effort, @erlingrj! I only left some basic code hygiene remarks. I noticed there are quite a few `FIXME` lingering in the code. I know this might be an annoying request, but if possible, please link them to a GitHub issue for tracking...

Thanks for the feedback. The PR is not quite ready, I marked it ready to be able to run the full test suite. I will address your comments and go through the FIXMEs and everything. I will request a review when I am done with that.

File: .github/workflows/c-zephyr-tests.yml
Comment (erlingrj) [reply .github/workflows/c-zephyr-tests.yml:44]: The thought was to run each test category to completion before starting the next. I had some cases where it run for 40m before returning the error which could have been caught and reported in 5m if we did each test-category to completion.

File: core/src/main/java/org/lflang/generator/EnclaveInfo.java
Comment (erlingrj) [reply core/src/main/java/org/lflang/generator/EnclaveInfo.java:3]: AFAIK the fields of a record are immutable, I would like to change the fields of the enclaveInfo as we explore the reactor-graph

File: core/src/main/java/org/lflang/generator/c/CActionGenerator.java
Comment (erlingrj) [reply core/src/main/java/org/lflang/generator/c/CActionGenerator.java:1]: I tried removing it

File: core/src/main/java/org/lflang/generator/c/CReactionGenerator.java
Comment (erlingrj) [reply core/src/main/java/org/lflang/generator/c/CReactionGenerator.java:330]: It is in CEnclavedReactorTransformation

**Review by erlingrj — COMMENTED** [review]
I have added more documents and fixed some FIXMEs

**Comment by erlingrj** [comment]
I have tagged @petervdonovan @edwardalee and @lhstrh for reviewing this and the reactor-c PR. We might not need a review from everyone but I think you are the ones in the best position to help me get these PR into an acceptable state.

**Comment by cmnrd** [comment]
I converted this PR to draft so that it uses less CI jobs. I hope that is Ok. Feel free to switch back to "ready for review" when this is ready, and you need a full CI run.

**Comment by erlingrj** [comment]
@petervdonovan @lhstrh @edwardalee I have now updated the tracing API also to deal with multiple environments. This PR and  https://github.com/lf-lang/reactor-c/pull/212 are now ready for review

File: core/src/main/java/org/lflang/generator/c/CTriggerObjectsGenerator.java
Comment (petervdonovan) [inline core/src/main/java/org/lflang/generator/c/CTriggerObjectsGenerator.java:130]: I do not totally understand the FIXMEs around this method. In any case we can just remember to open issues for them when this is merged.

File: core/src/main/java/org/lflang/generator/c/CUtil.java
Comment (petervdonovan) [inline core/src/main/java/org/lflang/generator/c/CUtil.java:811]: Probably all of these method-level comments should be JavaDoc (`/** */`)

File: core/src/main/java/org/lflang/generator/python/PythonGenerator.java
Comment (petervdonovan) [reply core/src/main/java/org/lflang/generator/python/PythonGenerator.java:114]: Why was this marked as resolved?

**Review by petervdonovan — APPROVED** [review]
I did not understand everything going on here, but I did not spot any important issues.

File: test/C/c/sendreceive.c
Comment (erlingrj) [reply test/C/c/sendreceive.c:7]: Yes, this is because name collisions. I think that the next step here is to fix this problem by re-architecting the platform support files such that we can include "platform.h" without including e.g. "windows.h" and "pthread.h". 

I think this re-architecting should be a separate PR.

File: core/src/main/java/org/lflang/generator/c/CUtil.java
Comment (erlingrj) [reply core/src/main/java/org/lflang/generator/c/CUtil.java:811]: Yes, thanks!

**Comment by erlingrj** [comment]
Thanks for your feedback! I have addressed it now. I think @petervdonovan raises an important point about the pollution of the namespaces. I have an issue on this here: https://github.com/lf-lang/reactor-c/issues/237. I can commit to addressing this issue next. I would like to merge these two big PRs now and rather address the re-architecting of the platform-support in a separate PR.

What do you think?

**Comment by petervdonovan** [comment]
> What do you think?

I 100% agree. I do not want the namespace problems to hold up this PR and would be willing to help with them later

**Comment by erlingrj** [comment]
Lets merge this, @lhstrh ?

**Review by lhstrh — APPROVED** [review]
LGTM (there is a comment staged from an earlier look at this, but I don't know what it is, so we'll see as soon as I hit the "approve" button).

**Comment by erlingrj** [comment]
I dont have permissions to merge the branch or add it to the merge queue. I can only keep updating the branch which triggers a new run of the tests.

**Pull request merged**
